### PR TITLE
Add vinyl section to configuration reference

### LINKED ISSUE #4018: [Config] Document 'vinyl' configuration settings
**Product:** Tarantool
**Since:** 3.0
**Root document:** 
- A new `vinyl` section in the [Configuration reference](https://www.tarantool.io/en/doc/latest/reference/configuration/configuration_reference/).

**SME:** @ Totktonada

# Details

Options:
- https://github.com/tarantool/tarantool/blob/f58bfc97c627f1838696697339fb3e443c05767a/src/box/lua/config/instance_config.lua#L1168

Related topics:
- https://www.tarantool.io/en/doc/latest/reference/configuration/

### PR DESCRIPTION
**Pull request by lenkis**

### CODE REVIEW

File: doc/reference/configuration/configuration_reference.rst [hunk 1]
```diff
@@ -3824,6 +3824,219 @@ snapshot.by.*
     | Default: 10^18
     | Environment variable: TT_SNAPSHOT_BY_WAL_SIZE
 
+???
```
Comment (Totktonada) [inline doc/reference/configuration/configuration_reference.rst]: It seems, it is a work-in-progress marker that was forgotten when the job has been done.

File: doc/reference/configuration/configuration_reference.rst [hunk 2]
```diff
@@ -3824,6 +3824,219 @@ snapshot.by.*
...
+    ``vinyl`` can be defined in any :ref:`scope <configuration_scopes>`.
+
+-   :ref:`vinyl.bloom_fpr <configuration_reference_vinyl_bloom_fpr>`
+-   :ref:`vinyl.cache <configuration_reference_vinyl_cache>`
+-   :ref:`vinyl.defer_deletes <configuration_reference_vinyl_defer_deletes>`
+-   :ref:`vinyl.dir <configuration_reference_vinyl_dir>`
+-   :ref:`vinyl.max_tuple_size <configuration_reference_vinyl_max_tuple_size>`
+-   :ref:`vinyl.memory <configuration_reference_vinyl_memory>`
+-   :ref:`vinyl.page_size <configuration_reference_vinyl_page_size>`
+-   :ref:`vinyl.range_size <configuration_reference_vinyl_>`
```
Comment (Totktonada) [inline doc/reference/configuration/configuration_reference.rst]: This link [is not rendered](https://docs.d.tarantool.io/en/doc/gh-4018-vinyl-config/reference/configuration/configuration_reference/#vinyl) on the website.

File: doc/reference/configuration/configuration_reference.rst [hunk 3]
```diff
@@ -3824,6 +3824,219 @@ snapshot.by.*
...
+
+    A directory where vinyl files or subdirectories will be stored.
+
+    This option may contain a relative file path.
+    In this case, it is interpreted as relative to
+    :ref:`process.work_dir <configuration_reference_process_work_dir>`.
+
+    |
+    | Type: string
+    | Default: var/lib/{{ instance_name }}'
```
Comment (Totktonada) [inline doc/reference/configuration/configuration_reference.rst]: Unbalanced single quote. As I see, other directories with a template variable are quoted, so I would do it here as well for consistency.

File: doc/reference/configuration/configuration_reference.rst [hunk 4]
```diff
@@ -3824,6 +3824,219 @@ snapshot.by.*
...
+
+..  _configuration_reference_vinyl_memory:
+
+..  confval:: vinyl.memory
+
+    The maximum number of in-memory bytes that vinyl uses.
+
+    |
+    | Type: integer
+    | Default: 28 * 1024 * 1024
```
Comment (Totktonada) [inline doc/reference/configuration/configuration_reference.rst]: Typo: 28 -> 128.

File: doc/reference/configuration/configuration_reference.rst [hunk 5]
```diff
@@ -3824,6 +3824,219 @@ snapshot.by.*
...
+..  confval:: vinyl.bloom_fpr
+
+    Bloom filter false positive rate -- the suitable probability of the
+    `bloom filter <https://en.wikipedia.org/wiki/Bloom_filter>`_
+    to give a wrong result.
+    The ``vinyl.bloom_fpr`` setting is a default value for one of the
+    options in the :ref:`Options for space_object:create_index() <box_space-create_index>` chart.
+
+    |
+    | Type: float
```
Comment (Totktonada) [inline doc/reference/configuration/configuration_reference.rst]: We call this type `number` across the document.

`float` may be confusing here, because sometimes it means a 32-bit floating point value (in contrast to 64-bit floating point value called `double`).

File: doc/reference/configuration/configuration_reference.rst [hunk 6]
```diff
@@ -3824,6 +3824,219 @@ snapshot.by.*
...
+..  _configuration_reference_vinyl_run_size_ratio:
+
+..  confval:: vinyl.run_size_ratio
+
+    Ratio between the sizes of different levels in the LSM tree.
+    The ``vinyl.run_size_ratio`` setting is a default value for one of the
+    options in the :ref:`Options for space_object:create_index() <box_space-create_index>` chart.
+
+    |
+    | Type: float
```
Comment (Totktonada) [inline doc/reference/configuration/configuration_reference.rst]: Same here: we call this type `number`.

File: doc/reference/configuration/configuration_reference.rst [hunk 7]
```diff
@@ -3824,6 +3824,219 @@ snapshot.by.*
...
+    :ref:`Options for space_object:create_index() <box_space-create_index>` chart.
+
+    If ``vinyl.range_size`` is nil or 0, and ``range_size`` is not specified
+    when the index is created, then Tarantool sets a value later depending on
+    performance considerations. To see the actual value, use
+    :doc:`index_object:stat().range_size </reference/reference_lua/box_index/stat>`.
+
+    |
+    | Type: integer
+    | Default: box.NULL
```
Comment (Totktonada) [inline doc/reference/configuration/configuration_reference.rst]: A reader can guess that a default behavior is described in this 'nil or 0' paragraph. However, the default is `box.NULL` and technically such a case is not described in the option description.

Moreover, a reader likely works with this machinery by creating/editing a YAML file, where we have neither `nil`, nor `box.NULL` (but we have a missed value and `null`).

I think we should make a little stop here and look how the options, which has `box_cfg` annotation in the schema, works.

----

First, a YAML file (or YAML document from a config storage) is parsed into Lua. If there is no option in YAML, it is `nil` in Lua. If the option is `null` in YAML, it is `box.NULL` in Lua.

| config.yaml | no option | `null`     |
| ----------- | --------- | ---------- |
| Lua         | `nil`     | `box.NULL` |

Next, default values are applied. Let's assume that the `default` annotation contains `box.NULL`.

| config.yaml   | no option  | `null`     |
| ----------- | --------- | ---------- |
| Lua           | `nil`      | `box.NULL` |
| apply default | `box.NULL` | `box.NULL` |

The important fact here is that a missing option is the same as `null` option.

Then the option is passed to the `box.cfg()` call. The `box.cfg()` code assumes that `box.NULL` and `''` (an empty string) means a default value (from the `default_cfg` table in `src/box/lua/load_cfg.lua`).

The important fact here is that `default = box.NULL` for an option with the `box_cfg` annotation means that the effective default is determined somewhere under the hood.

Usually `instance_config.lua` has `box.NULL` default for options, whose default is either hard to determine from Lua (depends on the OS, for example) or if the default is determined in runtime.

`load_cfg.lua::default_cfg.vinyl_range_size` is `nil`. Actually, it means that the effective default is determined in runtime.

Going down by the code we can see that `vinyl_range_size = nil` is represented as 0 in the C code and it means that a missed option, `null` and 0 in YAML are equivalent.

Going further, we can see that if the option is 0 in C, some heuristic is applied (`vy_lsm_range_size` in `src/box/vy_lsm.c`).

----

To sum up:

1. `box.NULL` is a marker that means that the effective default is determined in runtime. `config:get()` returns `box.NULL`. An effective value is something that `vy_lsm_range_size()` returns.

   It is a bit different from the existing `Default: box.NULL`, as for me. I would describe it somehow like `Default: box.NULL (means that an effective default is determined in runtime)`.
3. Missed option, `null` and 0 are equivalent in the case.
4. `nil` may confuse a reader of this document, because the document is likely read in context of YAML.

File: doc/reference/configuration/configuration_reference.rst [hunk 8]
```diff
@@ -3824,6 +3824,219 @@ snapshot.by.*
...
+    | Environment variable: TT_VINYL_DIR
+
+..  _configuration_reference_vinyl_max_tuple_size:
+
+..  confval:: vinyl.max_tuple_size
+
+    Size of the largest allocation unit,
+    for the vinyl storage engine. It can be increased if it
+    is necessary to store large tuples.
+    See also: :ref:`memtx.max_tuple_size <configuration_reference_memtx_max_tuple_size>`.
```
Comment (Totktonada) [inline doc/reference/configuration/configuration_reference.rst]: 1. I would place the 'See also' in its own paragraph.
2. The link is not rendered on the website.

File: doc/reference/configuration/configuration_reference.rst [hunk 9]
```diff
@@ -3824,6 +3824,219 @@ snapshot.by.*
...
+    | Environment variable: TT_VINYL_RUN_SIZE_RATIO
+
+..  _configuration_reference_vinyl_timeout:
+
+..  confval:: vinyl.timeout
+
+    The vinyl storage engine has a scheduler which does compaction.
+    When vinyl is low on available memory, the compaction scheduler
+    may be unable to keep up with incoming update requests.
+    In that situation, queries may time out after ``vinyl_timeout`` seconds.
```
Comment (Totktonada) [inline doc/reference/configuration/configuration_reference.rst]: Typo: `vinyl_timeout` -> `vinyl.timeout`.

**Review by Totktonada — APPROVED** [review]
Thank you for the new section!

I've no objections at whole, approved the PR.

I left comments related to the YAML configuration machinery.

However, I can't proofread parts that describe the vinyl machinery. You may possibly want a glance from one of developers who are working on the engine.

----

NB: A built reference page is [here](https://docs.d.tarantool.io/en/doc/gh-4018-vinyl-config/reference/configuration/configuration_reference/#vinyl).

File: doc/reference/configuration/configuration_reference.rst (hunk 7 above)
Comment (lenkis) [reply doc/reference/configuration/configuration_reference.rst]: Thanks a lot for the details! My suggestion, please say if it's ok:

```
    <...>
    If ``vinyl.range_size`` is specified (but the value is not ``null`` or 0), then
    it is used as the default value for the ``range_size`` option in the
    :ref:`Options for space_object:create_index() <box_space-create_index>` chart.

    If ``vinyl.range_size`` is not specified (or is explicitly set to ``null`` or 0),
    and ``range_size`` is not specified when the index is created,
    then Tarantool sets a value later depending on performance considerations.
    To see the actual value, use
    :doc:`index_object:stat().range_size </reference/reference_lua/box_index/stat>`.

    |
    | Type: integer
    | Default: box.NULL (means that an effective default is determined in runtime)
    | Environment variable: TT_VINYL_RANGE_SIZE
```

Comment (Totktonada) [reply doc/reference/configuration/configuration_reference.rst]: Seems OK, thanks for the update!

File: doc/reference/configuration/configuration_reference.rst [hunk 10]
```diff
@@ -3824,6 +3824,218 @@ snapshot.by.*
...
+    | Type: integer
+    | Default: 8 * 1024
+    | Environment variable: TT_VINYL_PAGE_SIZE
+
+..  _configuration_reference_vinyl_range_size:
+
+..  confval:: vinyl.range_size
+
+    The default maximum range size for a vinyl index, in bytes.
+    The maximum range size affects the decision whether to
```
Comment (andreyaksenov) [inline doc/reference/configuration/configuration_reference.rst]:
```suggestion
    The maximum range size affects the decision of whether to
```

File: doc/reference/configuration/configuration_reference.rst [hunk 11]
```diff
@@ -3824,6 +3824,218 @@ snapshot.by.*
...
+    |
+    | Type: integer
+    | Default: 1
+    | Environment variable: TT_VINYL_READ_THREADS
+
+..  _configuration_reference_vinyl_run_count_per_level:
+
+..  confval:: vinyl.run_count_per_level
+
+    The maximal number of runs per level in vinyl LSM tree.
```
Comment (andreyaksenov) [inline doc/reference/configuration/configuration_reference.rst]:
```suggestion
    The maximum number of runs per level in the vinyl LSM tree.
```

File: doc/reference/configuration/configuration_reference.rst [hunk 12]
```diff
@@ -3824,6 +3824,218 @@ snapshot.by.*
...
+    |
+    | Type: integer
+    | Default: 2
+    | Environment variable: TT_VINYL_RUN_COUNT_PER_LEVEL
+
+..  _configuration_reference_vinyl_run_size_ratio:
+
+..  confval:: vinyl.run_size_ratio
+
+    Ratio between the sizes of different levels in the LSM tree.
```
Comment (andreyaksenov) [inline doc/reference/configuration/configuration_reference.rst]:
```suggestion
    The ratio between the sizes of different levels in the LSM tree.
```

File: doc/reference/configuration/configuration_reference.rst [hunk 13]
```diff
@@ -3824,6 +3824,218 @@ snapshot.by.*
...
+    |
+    | Type: number
+    | Default: 3.5
+    | Environment variable: TT_VINYL_RUN_SIZE_RATIO
+
+..  _configuration_reference_vinyl_timeout:
+
+..  confval:: vinyl.timeout
+
+    The vinyl storage engine has a scheduler which does compaction.
```
Comment (andreyaksenov) [inline doc/reference/configuration/configuration_reference.rst]:
```suggestion
    The vinyl storage engine has a scheduler that performs compaction.
```

File: doc/reference/configuration/configuration_reference.rst [hunk 14]
```diff
@@ -3824,6 +3824,218 @@ snapshot.by.*
...
+-   :ref:`vinyl.run_count_per_level <configuration_reference_vinyl_run_count_per_level>`
+-   :ref:`vinyl.run_size_ratio <configuration_reference_vinyl_run_size_ratio>`
+-   :ref:`vinyl.timeout <configuration_reference_vinyl_timeout>`
+-   :ref:`vinyl.write_threads <configuration_reference_vinyl_write_threads>`
+
+..  _configuration_reference_vinyl_bloom_fpr:
+
+..  confval:: vinyl.bloom_fpr
+
+    Bloom filter false positive rate -- the suitable probability of the
```
Comment (andreyaksenov) [inline doc/reference/configuration/configuration_reference.rst]:
```suggestion
    A bloom filter's false positive rate -- the suitable probability of the
```

File: doc/reference/configuration/configuration_reference.rst [hunk 15]
```diff
@@ -3824,6 +3824,218 @@ snapshot.by.*
...
+    `bloom filter <https://en.wikipedia.org/wiki/Bloom_filter>`_
+    to give a wrong result.
+    The ``vinyl.bloom_fpr`` setting is a default value for one of the
+    options in the :ref:`Options for space_object:create_index() <box_space-create_index>` chart.
```
Comment (andreyaksenov) [inline doc/reference/configuration/configuration_reference.rst]: > for one of the options in the :ref:`Options for space_object:create_index() <box_space-create_index>` chart

Let's update the default value for this option from the old `vinyl_bloom_fpr` box.cfg option to a new `vinyl.bloom_fpr` config option. The same can be done for other vinyl options passed to `create_index()`.

File: doc/reference/configuration/configuration_reference.rst [hunk 16]
```diff
@@ -3824,6 +3824,218 @@ snapshot.by.*
...
+    |
+    | Type: string
+    | Default: 'var/lib/{{ instance_name }}'
+    | Environment variable: TT_VINYL_DIR
+
+..  _configuration_reference_vinyl_max_tuple_size:
+
+..  confval:: vinyl.max_tuple_size
+
+    Size of the largest allocation unit,
```
Comment (andreyaksenov) [inline doc/reference/configuration/configuration_reference.rst]:
```suggestion
    The size of the largest allocation unit
```

File: doc/reference/configuration/configuration_reference.rst [hunk 17]
```diff
@@ -3824,6 +3824,218 @@ snapshot.by.*
...
+    |
+    | Type: integer
+    | Default: 128 * 1024 * 1024
+    | Environment variable: TT_VINYL_MEMORY
+
+..  _configuration_reference_vinyl_page_size:
+
+..  confval:: vinyl.page_size
+
+    Page size. Page is a read/write unit for vinyl disk operations.
```
Comment (andreyaksenov) [inline doc/reference/configuration/configuration_reference.rst]:
```suggestion
    The page size. A page is a read/write unit for vinyl disk operations.
```

File: doc/reference/configuration/configuration_reference.rst [hunk 18]
```diff
@@ -3824,6 +3824,218 @@ snapshot.by.*
...
+    |
+    | Type: integer
+    | Default: box.NULL (means that an effective default is determined in runtime)
+    | Environment variable: TT_VINYL_RANGE_SIZE
+
+..  _configuration_reference_vinyl_read_threads:
+
+..  confval:: vinyl.read_threads
+
+    The maximum number of read threads that vinyl can use for some
```
Comment (andreyaksenov) [inline doc/reference/configuration/configuration_reference.rst]:
```suggestion
    The maximum number of read threads that vinyl can use for
```

File: doc/reference/configuration/configuration_reference.rst [hunk 19]
```diff
@@ -3824,6 +3824,218 @@ snapshot.by.*
...
+
+..  _configuration_reference_vinyl_timeout:
+
+..  confval:: vinyl.timeout
+
+    The vinyl storage engine has a scheduler which does compaction.
+    When vinyl is low on available memory, the compaction scheduler
+    may be unable to keep up with incoming update requests.
+    In that situation, queries may time out after ``vinyl.timeout`` seconds.
+    This should rarely occur, since normally vinyl
```
Comment (andreyaksenov) [inline doc/reference/configuration/configuration_reference.rst:4019]:
```suggestion
    This should rarely occur, since, normally vinyl
```

File: doc/reference/configuration/configuration_reference.rst [hunk 20]
```diff
@@ -3824,6 +3824,218 @@ snapshot.by.*
...
+    would throttle inserts when it is running low on compaction bandwidth.
```
Comment (andreyaksenov) [inline doc/reference/configuration/configuration_reference.rst]:
```suggestion
    throttles inserts when it is running low on compaction bandwidth.
```

File: doc/reference/configuration/configuration_reference.rst [hunk 21]
```diff
@@ -3824,6 +3824,218 @@ snapshot.by.*
...
+    Compaction can also be ordered manually with
```
Comment (andreyaksenov) [inline doc/reference/configuration/configuration_reference.rst]:
```suggestion
    Compaction can also be initiated manually with
```

File: doc/reference/configuration/configuration_reference.rst [hunk 22]
```diff
@@ -3824,6 +3824,218 @@ snapshot.by.*
...
+    |
+    | Type: integer
+    | Default: 60
+    | Environment variable: TT_VINYL_TIMEOUT
+
+..  _configuration_reference_vinyl_write_threads:
+
+..  confval:: vinyl.write_threads
+
+    The maximum number of write threads that vinyl can use for some
```
Comment (andreyaksenov) [inline doc/reference/configuration/configuration_reference.rst:4033]:
```suggestion
    The maximum number of write threads that vinyl can use for
```

File: doc/reference/configuration/configuration_reference.rst (hunk 15 above)
Comment (andreyaksenov) [inline doc/reference/configuration/configuration_reference.rst]:
```suggestion
    The ``vinyl.bloom_fpr`` setting is a default value for the ``bloom_fpr``
    option passed to the :ref:`space_object:create_index() <box_space-create_index>` method.
```

**Review by andreyaksenov — CHANGES_REQUESTED** [review]

File: doc/reference/configuration/configuration_reference.rst (hunk 15 above)
Comment (Totktonada) [reply doc/reference/configuration/configuration_reference.rst]: It is 0.05 in the box.cfg reference too, [as I see](https://www.tarantool.io/en/doc/latest/reference/configuration/#cfg-storage-vinyl-bloom-fpr).

**Review by andreyaksenov — APPROVED** [review]
